File: extractors/normalized_record.py
"""
Pydantic mirror of normalized-record.schema.json.

Every Python extractor imports NormalizedRecord from here and constructs one
per source record before writing to Postgres. This is what makes a new
extractor "additive" instead of risky — it can't drift from the shared
envelope because the type checker/validator enforces it at construction time,
not at report-render time (which is where the original crash-report bug
would have been caught, too late, if it had been caught at all).

Keep this in lockstep with:
  - contracts/normalized-record.schema.json  (source of truth)
  - contracts/normalizedRecord.ts            (TypeScript/Zod mirror)
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict


class SourceType(str, Enum):
    CRASH_REPORT = "crash_report"
    SIRI_FEEDBACK = "siri_feedback"
    SFA_ANALYTICS = "sfa_analytics"
    XP_AMP_TELEMETRY = "xp_amp_telemetry"
    SAFARI_HISTORY = "safari_history"
    SMS_ATTACHMENT = "sms_attachment"
    NETWORK_USAGE = "network_usage"
    GCLOUD_LOG = "gcloud_log"
    SYSLOG_LINE = "syslog_line"


class NormalizedRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    incident_id: str | None = None
    source_type: SourceType
    event_time: datetime | None = None
    bug_type: str | None = None
    process_name: str | None = None
    pid: int | None = None
    bundle_id: str | None = None
    fields: dict[str, Any] = {}
    raw_ref: str | None = None  # file_hash back-reference into ingested_files

File: extractors/normalizedRecord.ts
/**
 * Zod mirror of normalized-record.schema.json.
 *
 * Used by orchestrator/ and reporting/ (and any future TS extractor) so a
 * record read out of Postgres — or about to be written into it — is
 * validated against the exact same envelope Python extractors build against.
 *
 * Keep this in lockstep with:
 *   - contracts/normalized-record.schema.json  (source of truth)
 *   - contracts/normalized_record.py           (Python/Pydantic mirror)
 */

import { z } from "zod";

export const SourceType = z.enum([
  "crash_report",
  "siri_feedback",
  "sfa_analytics",
  "xp_amp_telemetry",
  "safari_history",
  "sms_attachment",
  "network_usage",
  "gcloud_log",
  "syslog_line",
]);
export type SourceType = z.infer<typeof SourceType>;

export const NormalizedRecord = z
  .object({
    incident_id: z.string().nullable().default(null),
    source_type: SourceType,
    event_time: z.string().datetime().nullable().default(null),
    bug_type: z.string().nullable().default(null),
    process_name: z.string().nullable().default(null),
    pid: z.number().int().nullable().default(null),
    bundle_id: z.string().nullable().default(null),
    fields: z.record(z.string(), z.unknown()).default({}),
    raw_ref: z.string().nullable().default(null),
  })
  .strict();

export type NormalizedRecord = z.infer<typeof NormalizedRecord>;
